src/codegen/unused_variable.rs

```diff
@@ -0,0 +1,99 @@
+use crate::codegen::cfg::{ControlFlowGraph, Vartable};
+use crate::parser::pt;
+use crate::sema::ast::{Expression, Function, Namespace};
+use crate::sema::symtable::VariableUsage;
+
+/// This struct saves the parameters to call 'check_side_effects_expressions'
+/// using 'expression.recurse'
+pub struct SideEffectsCheckParameters<'a> {
+    pub cfg: &'a mut ControlFlowGraph,
+    pub contract_no: usize,
+    pub func: Option<&'a Function>,
+    pub ns: &'a Namespace,
+    pub vartab: &'a mut Vartable,
+}
+
+/// Check if we should remove an assignment. The expression in the argument is the left-hand side
+/// of the assignment
+pub fn should_remove_assignment(ns: &Namespace, exp: &Expression, func: &Function) -> bool {
+    match &exp {
+        Expression::StorageVariable(_, _, contract_no, offset) => {
+            let var = &ns.contracts[*contract_no].variables[*offset];
+            if matches!(
```

Comment:
var.read is set to true for public/external variables already; this check should be redundant (i.e. remove line 22 to 27)